Implement selection I/O with type conversion

fortnern opened the pull request:
Also enables collective I/O with type conversion. Also added H5Pset_selection_io(), H5Pget_selection_io(), and H5Pget_no_selection_io_cause() API functions to manage the selection I/O feature, and H5Pset_modify_write_buf() and H5Pget_modify_write_buf() to allow the library to modify write buffers.

fortnern committed:
Initial implementation of selection I/O with type conversion.  Allows
Parallel collective I/O with type conversion, as long as selection I/O
is enabled.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Fix spelling

fortnern committed:
Merge branch 'select_io_tconv' of github.com:fortnern/hdf5 into select_io_tconv

fortnern committed:
Merge pull request #2418 from fortnern/select_io_tconv

Initial implementation of selection I/O with type conversion

Commit:
Add tests for selection I/O.

github-actions[bot] committed:
Committing clang-format changes

Commit:
Minor changes.

Commit:
Merge branch 'my_feature_select_io_tconv' of https://github.com/vchoi-hdfgroup/hdf5 into my_feature_select_io_tconv

Commit:
Correct spelling errors.

fortnern committed:
Merge pull request #2439 from vchoi-hdfgroup/my_feature_select_io_tconv

My feature select io tconv

fortnern committed:
Merge pull request #15 from HDFGroup/feature/select_io_tconv

Merge from canonical

Commit:
(1) Add 3 tests for multi-datasets
(2) Modify tests to make sure they cover type conversion for writing
(3) Fail the checking of collective I/O mode

github-actions[bot] committed:
Committing clang-format changes

Commit:
Correct spelling error.

Commit:
Merge branch '2my_feature_select_io_tconv' of https://github.com/vchoi-hdfgroup/hdf5 into 2my_feature_select_io_tconv

Commit:
Add one more test for multi-dataset.
Modify/refactor existing tests.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge pull request #2491 from vchoi-hdfgroup/2my_feature_select_io_tconv

(1) Add 3 tests for multi-datasets

fortnern committed:
Merge pull request #16 from HDFGroup/feature/select_io_tconv

Pull from canonical branch

fortnern committed:
Add actual I/O mode reporting to selection I/O pathway. Other minor
fixes.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge pull request #2508 from fortnern/select_io_tconv

Add actual I/O mode reporting to selection I/O pathway

fortnern committed:
Only report collective for actual I/O mode if we're actually doing
collective

fortnern committed:
Merge branch 'select_io_tconv' of github.com:fortnern/hdf5 into select_io_tconv

fortnern committed:
Merge pull request #2509 from fortnern/select_io_tconv

Only report collective for actual I/O mode if we're actually doing collective I/O

vchoi-hdfgroup committed:
Merge pull request #71 from HDFGroup/feature/select_io_tconv

Feature/select io tconv

Commit:
1) Add H5Pset/get_selection_io() API and associated changes: typedef enum, H5CX package changes etc.
2) Remove H5_use_selection_io_g global and H5_USE_SELECTION_IO environment variable
3) Modify tests to use the new API to turn on selection I/O
4) Add basic test for the new API
5) Correct spelling errors

github-actions[bot] committed:
Committing clang-format changes

Commit:
Correct spelling errors.

vchoi-hdfgroup committed:
Merge branch '2my_feature_select_io_tconv' of https://github.com/vchoi-hdfgroup/hdf5 into 2my_feature_select_io_tconv

vchoi-hdfgroup committed:
Add multi dataset test that will trigger write/read operations with one single multi-write/read for datasets:
--with no type conversion
--with type conversion but no background buffer
--with type conversion and background buffer

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge pull request #2573 from vchoi-hdfgroup/2my_feature_select_io_tconv

2my feature select io tconv

fortnern committed:
Merge pull request #17 from HDFGroup/feature/select_io_tconv

Merge from canonical

fortnern committed:
Add resons for no selection I/O to "no collective cause" reporting.  Add
automatic selection I/O enablind when it is beneficial.  Other
refactoring/fixes.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge pull request #2659 from fortnern/select_io_tconv

Selection I/O improvements

vchoi-hdfgroup committed:
Merge pull request #72 from HDFGroup/feature/select_io_tconv

Feature/select io tconv

fortnern committed:
Change background buffer to be global for multi dataset operations.
Allow user-supplied background buffer for multi dataset operations.
Switch background buffer allocation from calloc to malloc.

fortnern committed:
Merge branch 'select_io_tconv' of github.com:fortnern/hdf5 into select_io_tconv

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Spelling

fortnern committed:
Merge pull request #2716 from fortnern/select_io_tconv

Background buffer changes

vchoi-hdfgroup committed:
Merge pull request #73 from vchoi-hdfgroup/feature/select_io_tconv

Feature/select io tconv

vchoi-hdfgroup committed:
Changes for adding H5Pget_no_selection_io_cause() API.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge pull request #2796 from vchoi-hdfgroup/2my_feature_select_io_tconv

2my feature select io tconv

fortnern committed:
Merge pull request #18 from HDFGroup/feature/select_io_tconv

Feature/select io tconv

github-actions[bot] committed:
Committing clang-format changes

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Merge pull request #74 from vchoi-hdfgroup/feature/select_io_tconv

Feature/select io tconv

fortnern committed:
Implement in-place type conversion for read with selection I/O

fortnern committed:
Merge branch 'select_io_tconv' of github.com:fortnern/hdf5 into select_io_tconv

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Implement H5D_SEL_IO_NO_VECTOR_OR_SELECTION_IO_CB

fortnern committed:
Fix typo

fortnern committed:
Merge pull request #2805 from fortnern/select_io_tconv

In-place type conversion for read with selection I/O

vchoi-hdfgroup committed:
Merge pull request #75 from vchoi-hdfgroup/feature/select_io_tconv

Feature/select io tconv

fortnern committed:
Enable type conversion with selection I/O in serial.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge pull request #2814 from fortnern/select_io_tconv

Enabled selection I/O with type conversion in serial

fortnern committed:
Implement H5Pset_modify_write_buf and H5Pget_modify_write_buf

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Doc updates for selection I/O API functions

fortnern committed:
Fix typos

fortnern committed:
Merge pull request #2820 from fortnern/select_io_tconv

Implement H5Pset/get_modify_write_buf()

fortnern committed:
Merge branch 'develop' into select_io_tconv

fortnern committed:
Merge pull request #2821 from fortnern/select_io_tconv

Bring in sync with develop

fortnern committed:
Add RELEASE.txt notes for selection I/O and modify write buffer
features.

fortnern committed:
Merge pull request #2822 from fortnern/select_io_tconv

Add RELEASE.txt notes for selection I/O and modify write buffer features

fortnern committed:
Fix issues with serial and add documentation

fortnern committed:
Typo

fortnern committed:
Merge pull request #2824 from fortnern/select_io_tconv

Fixes for serial

byrnHDF commented on test/select_io_dset.c:
```diff
@@ -0,0 +1,3212 @@
+#include "testhdf5.h"
+#include "H5srcdir.h"
+
+const char *FILENAME[] = {"select_io", /* 0 */
```
Is this file created and deleted - or does it need to be added to the build system cleanup sections?
(Also same for parallel test)

fortnern replied:
It looks like the file is deleted by h5_cleanup() in the serial test and MPI_File_delete() in the parallel test

brtnfld commented on src/H5Ppublic.h:
```diff
@@ -8204,6 +8265,190 @@ H5_DLL herr_t H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned ran
+ *
+ * \brief Retrieves the selection I/O mode
+ *
+ * \param[in] plist_id Property list identifier
```
Should be dxpl_id, but I think it should use the alias \plist_id{dxpl_id}

But, what is the convention? Some APIs use dxpl_id and some used plist_id, and yet they all have the same description. I think all of the new ones should use plist_id.

fortnern replied:
Updated, thanks. Not sure on the convention.

fortnern committed:
Update API parameter names and add tracing

fortnern committed:
Merge pull request #2825 from fortnern/select_io_tconv

Update API parameter names and add tracing

vchoi-hdfgroup commented on src/H5Ppublic.h:
```diff
@@ -8204,6 +8265,190 @@ H5_DLL herr_t H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned ran
+ *          Selection I/O was not performed because the selection I/O mode is DEFAULT and the library chose it
+ * to be off for this case
+ *
+ * \since 1.14.3
```
Should be 1.14.1

brtnfld commented on src/H5Ppublic.h:
```diff
@@ -8204,6 +8265,190 @@ H5_DLL herr_t H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned ran
+ *
+ * \dxpl_id{plist_id}
+ * \param[out] no_selection_io_cause A bitwise set value indicating the relevant
+ *                                   causes that prevented selection I/O to be performed
```
to be -> from being

brtnfld commented on src/H5Ppublic.h:
```diff
@@ -357,11 +357,72 @@ typedef enum H5D_mpio_no_collective_cause_t {
 } H5D_mpio_no_collective_cause_t;
 //! <!-- [H5D_mpio_no_collective_cause_t_snip] -->
 
+/**
```
Why are these not in a struct like the other no*cause cases? Why unsigned?

fortnern replied:
The others are in enums. We decided to avoid putting bitwise values in enums because that is inconsistent with the normal usage of enums and potentially confusing, since the flags parameter can take on a value that does not match one of the enum values. In addition, unless I'm missing something there's no advantage using enum values instead of simple #defines.

fortnern replied:
They are unsigned because they are bitflag values and need to be able to be combined in a bitwise fashion

fortnern replied:
We plan to change the other no*cause values to #defines with 2.0

brtnfld commented on src/H5Pdxpl.c:
```diff
@@ -2386,3 +2496,248 @@ H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned rank, H5S_seloper
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5Pget_no_selection_io_cause(hid_t plist_id, uint32_t *no_selection_io_cause /*out*/)
```
why is this marked` /*out*/ `and others are not?

bmribler commented on testpar/t_subfiling_vfd.c:
```diff
@@ -1011,12 +1014,19 @@ test_read_different_stripe_size(void)
     curr_nerrors = nerrors;
 
+    dxpl_id = H5Pcreate(H5P_DATASET_XFER);
+    VRFY((dxpl_id >= 0), "DCPL creation succeeded");
```
DXPL?  More below.

bmribler commented on testpar/t_subfiling_vfd.c:
```diff
@@ -427,6 +427,9 @@ test_stripe_sizes(void)
     VRFY((dxpl_id >= 0), "DCPL creation succeeded");
 
+    /* Set selection I/O mode on DXPL */
+    VRFY((H5Pset_selection_io(dxpl_id, H5D_SELECTION_IO_MODE_ON) >= 0), "H5Pset_selection_io succeeded");
```
There is a minor inconsistency in the messages.  Sometimes, "< function name > succeeded", and other times, "Doing something succeeded" or "doing something succeeded".

fortnern replied:
I think this is fine

bmribler commented on testpar/t_dset.c:
```diff
@@ -3538,34 +3555,59 @@ test_no_collective_cause_mode(int selection_mode)
+
+        /* Disable selection I/O since datatype conversion is supported in collective with selection I/O */
+        ret = H5Pset_selection_io(dxpl_write, H5D_SELECTION_IO_MODE_OFF);
+
```
Extra line?

fortnern replied:
You mean you think the comment should be broken into two lines? I generally rely on the formatter to pick line breaks, which I think is set at 110. This is 107.

bmribler replied:
No, I meant there is an extra line, 3574.  It's ok if it's intentional, hence, the question mark. :)

fortnern replied:
Ok I'll reformat thanks

bmribler commented on testpar/t_dset.c:
```diff
@@ -3348,13 +3348,26 @@ test_actual_io_mode(int selection_mode)
      * currently */
-    if (!H5_use_selection_io_g) {
+
+    dxpl_id = H5Pcreate(H5P_DATASET_XFER);
```
No verifying

bmribler commented on testpar/t_2Gio.c:
```diff
@@ -4072,6 +4094,20 @@ test_no_collective_cause_mode(int selection_mode)
                                           &no_collective_cause_global_write);
     VRFY((ret >= 0), "retrieving no collective cause succeeded");
 
+    ret = H5Pget_no_selection_io_cause(dxpl_write, &no_selection_io_cause_read);
```
no_selection_io_cause_write?

vchoi-hdfgroup committed:
Merge pull request #77 from vchoi-hdfgroup/feature/select_io_tconv

Feature/select io tconv

fortnern committed:
Address review comments

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge pull request #2829 from fortnern/select_io_tconv

Address review comments

fortnern committed:
Update DXPL table in doxygen

fortnern committed:
Merge pull request #2830 from fortnern/select_io_tconv

Update DXPL table in doxygen

vchoi-hdfgroup committed:
Merge pull request #78 from vchoi-hdfgroup/feature/select_io_tconv

Feature/select io tconv

vchoi-hdfgroup committed:
Merge pull request #79 from vchoi-hdfgroup/feature/select_io_tconv

Feature/select io tconv

vchoi-hdfgroup committed:
Add H5Pset_modify_write_buf() property to gen_plist.c to generate plist files.
Add H5Pset_modify_write_buf() to enc_dec_plist.c for testing.
Add the new plist files for dxpl.

fortnern committed:
Merge pull request #2840 from vchoi-hdfgroup/2my_feature_select_io_tconv

2my feature select io tconv

jhendersonHDF commented:
Just an overall question on this PR: is there a downside currently to always enabling selection I/O as opposed to needing to set it with the new API routines?

jhendersonHDF commented on test/select_io_dset.c:
```diff
@@ -0,0 +1,3212 @@
+        HDfree(s3_total_rbuf3);
+        HDfree(s4_total_wbuf3);
+        HDfree(s4_total_wbuf3_bak);
+        HDfree(s1_total_rbuf3);
```
All these buffers above should probably be set to NULL. Otherwise if the next iteration fails somewhere we might double-free segfault in the error cleanup section.

fortnern replied:
Fixed, thanks

jhendersonHDF commented on testpar/t_select_io_dset.c:
```diff
@@ -0,0 +1,3775 @@
+        HDfree(s3_total_rbuf3);
+        HDfree(s4_total_wbuf3);
+        HDfree(s4_total_wbuf3_bak);
+        HDfree(s1_total_rbuf3);
```
Suggest setting all these buffers to NULL to prevent double-free segfault.

jhendersonHDF commented on testpar/t_select_io_dset.c:
```diff
@@ -0,0 +1,3775 @@
+
+    /* Delete file */
+    if (MAINPROCESS)
+        if (MPI_File_delete(FILENAME, MPI_INFO_NULL) != MPI_SUCCESS)
```
Can this test be run with other parallel-enabled VFDs? If so this should probably be a call to H5Fdelete (either collectively with the file fapl or by MAINPROCESS with a serial fapl) so that VFDs like the Subfiling VFD can correctly delete the test file.

fortnern replied:
Fixed, thanks

jhendersonHDF commented on testpar/t_dset.c:
```diff
@@ -3538,34 +3556,59 @@ test_no_collective_cause_mode(int selection_mode)
-        no_collective_cause_global_expected |= H5D_MPIO_DATA_TRANSFORMS;
+
+        /* Set transform */
+        ret = H5Pset_data_transform(dxpl_write, "x+1");
```
As an aside, what's the current level of support for collective selection I/O in the library? There's an open issue about supporting data transformations with parallel compression and since parallel compressed writes require collective I/O, seems like we could at least partially support that now.

fortnern replied:
Selection I/O supports transforms with collective, but selection I/O doesn't support compressions, so you can't have both yet unfortunately

jhendersonHDF commented on src/H5CXprivate.h:
```diff
@@ -115,6 +115,9 @@ H5_DLL herr_t H5CX_get_filter_cb(H5Z_cb_t *filter_cb);
 H5_DLL herr_t H5CX_get_dt_conv_cb(H5T_conv_cb_t *cb_struct);
+H5_DLL herr_t H5CX_get_selection_io_mode(H5D_selection_io_mode_t *selection_io_mode);
+H5_DLL herr_t H5CX_get_no_selection_io_cause(uint32_t *no_selection_io_cause);
+H5_DLL herr_t H5CX_get_modify_write_buf(hbool_t *modify_write_buf);
```
No need for set routines for selection io mode or modify write buf?

fortnern replied:
Not needed by the library since the library never changes it. There are several other properties where this is the case. For something like this it's best not to create this routine since we really want to make sure we don't violate const if the app didn't set this property.

jhendersonHDF commented on src/H5Ppublic.h:
```diff
@@ -8204,6 +8265,191 @@ H5_DLL herr_t H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned ran
+ *
+ * \details H5Pset_selection_io() sets the selection I/O mode
+ *          \p selection_io_mode in the dataset transfer property
+ *          list \p dxpl_id.
```
Should this be updated to plist_id?

jhendersonHDF commented on src/H5Ppublic.h:
```diff
@@ -8204,6 +8265,191 @@ H5_DLL herr_t H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned ran
+ * \return \herr_t
+ *
+ * \details H5Pget_selection_io() queries the selection I/O mode set in
+ *          in the dataset transfer property list \p dxpl_id.
```
plist_id?

jhendersonHDF commented on src/H5Ppublic.h:
```diff
@@ -8204,6 +8265,191 @@ H5_DLL herr_t H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned ran
+ *
+ *          Values that can be set in \p selection_io_mode:
+ *          \snippet this H5D_selection_io_mode_t_snip
+ *          \click4more
```
Is this some special doxygen command?

fortnern replied:
It seems to print "(Click on a enumerator, field, or type for more information.)" in the html

jhendersonHDF commented on src/H5Ppublic.h:
```diff
@@ -8204,6 +8265,191 @@ H5_DLL herr_t H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned ran
+ *
+ * \details H5Pset_modify_write_buf() sets whether the library is allowed to
+ *          modify the contents of write buffers passed to HDF5 API routines
+ *          that are passed the dataset transfer property list \p dxpl_id.  The
```
plist_id?

jhendersonHDF commented on src/H5Ppublic.h:
```diff
@@ -8204,6 +8265,191 @@ H5_DLL herr_t H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned ran
+ * \return \herr_t
+ *
+ * \details H5Pget_modify_write_buf() gets the "modify write buffer" property
+ *          from the dataset transfer property list \p dxpl_id.  This property
```
plist_id?

jhendersonHDF commented on src/H5Ppublic.h:
```diff
@@ -8204,6 +8265,191 @@ H5_DLL herr_t H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned ran
+ *          from the dataset transfer property list \p dxpl_id.  This property
+ *          determines whether the library is allowed to  modify the contents of
+ *          write buffers passed to HDF5 API routines that are passed
+ *          \p dxpl_id.  The default value for modify_write_buf is FALSE.
```
plist_id?

jhendersonHDF commented on src/H5Pdxpl.c:
```diff
@@ -441,6 +461,24 @@ H5P__dxfr_reg_prop(H5P_genclass_t *pclass)
+        HGOTO_ERROR(H5E_PLIST, H5E_CANTINSERT, FAIL, "can't insert property into class")
+
+    /* Register the cause of no selection I/O property */
+    /* (Note: this property should not have an encode/decode callback -QAK) */
```
Remove copied comment from QAK

jhendersonHDF commented on src/H5Pdxpl.c:
```diff
@@ -2386,3 +2496,248 @@ H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned rank, H5S_seloper
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5Pget_selection_io(hid_t plist_id, H5D_selection_io_mode_t *selection_io_mode /*out*/)
```
Should check for H5P_DEFAULT and set plist_id to correct library default DXPL if that's passed in

fortnern replied:
Are you sure about this? None of the other H5Pget* functions seem to do this

fortnern replied:
I guess it might be useful to make this work, especially since we're bad about documenting the default values for properties (I just had to scour through the code to find some to fix one of the above tests), but we should really apply it to all at once as a separate PR

jhendersonHDF replied:
Yeah I think we should check for and handle H5P_DEFAULT in every API routine where a plist is passed in. I think it's just something that's easy to forget and so many routines are missing the check. Maybe we should move the logic down into H5P_object_verify since it accepts a "plist type" parameter. Other places where H5I_object is used won't be covered but it'd be a step in the right direction.

jhendersonHDF commented on src/H5Pdxpl.c:
```diff
@@ -2386,3 +2496,248 @@ H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned rank, H5S_seloper
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5Pget_no_selection_io_cause(hid_t plist_id, uint32_t *no_selection_io_cause /*out*/)
```
Should check for H5P_DEFAULT and set plist_id to correct library default DXPL if that's passed in

fortnern replied:
See above

jhendersonHDF commented on src/H5Pdxpl.c:
```diff
@@ -2386,3 +2496,248 @@ H5Pset_dataset_io_hyperslab_selection(hid_t plist_id, unsigned rank, H5S_seloper
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5Pget_modify_write_buf(hid_t plist_id, hbool_t *modify_write_buf /*out*/)
```
Should check for H5P_DEFAULT and set plist_id to correct library default DXPL if that's passed in

fortnern replied:
See above

fortnern commented:
> Just an overall question on this PR: is there a downside currently to always enabling selection I/O as opposed to needing to set it with the new API routines?

It could be slower for parallel I/O with strided selections since there's currently no compact way to pass a regular selection (only vector I/O is implemented). When used with type conversion it requires potentially very large type conversion buffers. And there are also intermittent failures in the parallel tests with MPICH (most likely an MPICH issue).

fortnern committed:
Address review comments

fortnern committed:
Merge pull request #2845 from fortnern/select_io_tconv

Address review comments

jhendersonHDF commented on src/H5Dio.c:
```diff
@@ -833,25 +910,41 @@ H5D__ioinfo_init(size_t count, H5D_dset_io_info_t *dset_info, H5D_io_info_t *io_
+    /* Check if we could potentially use in-place type conversion */
+    if (op_type == H5D_IO_OP_READ)
+        /* Always on for read (modulo other restrictions that are handled in layout callbacks) */
+        io_info->may_use_in_place_tconv = (op_type == H5D_IO_OP_READ);
```
Should this be hoisted out of the if block here since the conditional is the same as the check in the if? Though I guess the field gets initialized in the `op_type == H5D_IO_OP_READ` case anyway due to the memset of io_info above.

fortnern replied:
Oops, that logic was left over from before I implemented modify_write_buf. I'll fix it.

fortnern committed:
Address review comments

fortnern committed:
Merge pull request #2848 from fortnern/select_io_tconv

Address review comments

jhendersonHDF commented on src/H5Dpkg.h:
```diff
@@ -67,7 +67,49 @@
+                                        : (DINFO)->type_info.dst_type_size;                                  \
+                                                                                                             \
+            /* Make sure the memory type is not smaller than the file type, otherwise the memory buffer      \
+             * won't be big enough tp serve as the type conversion buffer */                                 \
```
tp -> to. Guess codespell didn't catch that one.

jhendersonHDF reviewed (approved):
I'm not as familiar with selection I/O as I'd like to be for reviewing this PR, but everything here looks reasonable enough to me.

fortnern committed:
Skip checking for in-place tconv if not doing selection I/O, fix typo

fortnern commented:
I'll fix the typo and have one minor change then I think it's good to go

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge pull request #2850 from fortnern/select_io_tconv

Skip checking for in-place tconv if not doing selection I/O, fix typo

fortnern commented:
Working on a couple tests that we discussed at the meeting, but should be fine to merge without if we're in a hurry. Edit: done.

fortnern committed:
Add tests to make sure modify_write_buf property enables
selection/collective I/O

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge pull request #2851 from fortnern/select_io_tconv

Add tests to make sure modify_write_buf property enables selection/collective I/O

fortnern committed:
Minor portability fix

fortnern committed:
Merge pull request #2852 from fortnern/select_io_tconv

Minor portability fix
